### PR TITLE
Feature/csv export

### PR DESCRIPTION
**Pull request by jonathoncove2**
Add a new service to export objects as CSV 
Added new AdminAuthorised controller to compile all pages and exporting using this service 
Added a dashboard in the 'content' area to trigger the export

### CODE REVIEW

File: src/JNCC.PublicWebsite.Core/Services/CSVExportService.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+					else
+					{
+						//Format certain values certain ways.
+						if (piType.IsEquivalentTo(typeof(DateTime)))
+						{
+							sbOutput.Append(FormatLine(((DateTime)propValue).ToString("dd.MM.yy")));
+						}
+						else if (piType.IsEquivalentTo(typeof(decimal)))
+						{
+							sbOutput.Append(FormatLine(((decimal)propValue).ToString("N2")));
```
Comment (NobleMule) [inline src/JNCC.PublicWebsite.Core/Services/CSVExportService.cs:110]: Is it right to always assume 2dp?

File: src/JNCC.PublicWebsite.Core/Services/CSVExportService.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+					if (propValue == null)
+					{
+						sbOutput.Append(FormatLine("NULL"));
+					}
+					else
+					{
+						//Format certain values certain ways.
+						if (piType.IsEquivalentTo(typeof(DateTime)))
+						{
+							sbOutput.Append(FormatLine(((DateTime)propValue).ToString("dd.MM.yy")));
```
Comment (NobleMule) [inline src/JNCC.PublicWebsite.Core/Services/CSVExportService.cs]: Is this their preferred date format? Or would ISO dates be better?

File: src/JNCC.PublicWebsite.Core/Services/CSVExportService.cs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+				sbOutput.Append("\r\n");
+			}
+
+			return sbOutput.ToString();
+		}
+
+		public static string FormatLine(string intput)
+		{
+			return "\"" + intput.Replace(",", "").Replace(System.Environment.NewLine, "").Replace("\t", "").Replace("\n", "").Replace("\r", "") + "\"" + ",";
```
Comment (NobleMule) [inline src/JNCC.PublicWebsite.Core/Services/CSVExportService.cs:127]: Why do we want to remove commas from a line? Is it possible that they're part of the value?

Incidentally, formatting a number with "N" will add commas to numbers over 1000

File: src/JNCC.PublicWebsite.Core/Controllers/ApiControllers/PageExportApiController.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            IUserService _userService = Services.UserService;
+            
+            var pages = _contentService.GetRootContent().SelectRecursive(c => c.Descendants());
+            var url = HttpContext.Current.Request.Url.Scheme + "://" + HttpContext.Current.Request.Url.GetHostAndPort();
+
+            var formatted = pages.Select(p => new PageExportModel(p, _userService, Umbraco, url));
+
+            CSVExportService.GenericExport(formatted.ToArray(), "Pages", true);
+
+            return InternalServerError(new Exception("There was an issue generating the report"));
```
Comment (NobleMule) [inline src/JNCC.PublicWebsite.Core/Controllers/ApiControllers/PageExportApiController.cs:36]: I presume the fact that a successful run ends the current response means that this error doesn't happen in a successful case?

File: src/JNCC.PublicWebsite.Core/Services/CSVExportService.cs (hunk 1 above)
Comment (jonathoncove2) [reply src/JNCC.PublicWebsite.Core/Services/CSVExportService.cs:110]: For decimals, this always seems reasonable

File: src/JNCC.PublicWebsite.Core/Controllers/ApiControllers/PageExportApiController.cs (hunk 4 above)
Comment (jonathoncove2) [reply src/JNCC.PublicWebsite.Core/Controllers/ApiControllers/PageExportApiController.cs:36]: Correct

File: src/JNCC.PublicWebsite.Core/Services/CSVExportService.cs (hunk 2 above)
Comment (jonathoncove2) [reply src/JNCC.PublicWebsite.Core/Services/CSVExportService.cs]: Good spot, adding a time in here would probably be useful

File: src/JNCC.PublicWebsite.Core/Services/CSVExportService.cs (hunk 3 above)
Comment (jonathoncove2) [reply src/JNCC.PublicWebsite.Core/Services/CSVExportService.cs:127]: This should really be called 'format cell' or something similar. If the value of the cell has a comma, it'll spoil the formatting. The alternative is to wrap the value in quotes, but this method is easier.

**Review by NobleMule — APPROVED** [review]

**Comment by jonathoncove2** [comment]
Closing this PR as it was incorrectly made from the master branch, new PR is https://github.com/jncc/jncc-website/pull/105

**Pull request closed without merging**